File: leetcode/0367-valid-perfect-square/solution.py
class Solution:
    def isPerfectSquare(self, num: int) -> bool:
        lo, hi = 1, num
        while lo < hi:
            mid = lo + (hi - lo) // 2
            
            if mid * mid == num:
                return True
            elif mid * mid < num:
                lo = mid + 1
            else:
                hi = mid - 1
        
        return lo * lo == num

File: leetcode/0367-valid-perfect-square/solution.cpp
class Solution {
public:
    bool isPerfectSquare(int num) {
        long long lo = 1, hi = num;
        while (lo < hi) {
            long long mid = lo + (hi - lo) / 2;
            
            if (mid * mid == num)
                return true;
            else if (mid * mid < num)
                lo = mid + 1;
            else
                hi = mid - 1;
        }
        
        return lo * lo == num;
    }
};